### PR TITLE
[core] Introduce non-bucket mode for append-only table

### PR DESCRIPTION
**Pull request by leaves12138**
[core] add negative bucket mode, separate compaction from appendonly writer, add test for negative bucket table #1224

In paimon core project, add the negative bucket process.

### Tests

paimon-core/src/test/java/org/apache/paimon/append/AppendOnlyTableCompactionTest.java
paimon-core/src/test/java/org/apache/paimon/table/sink/CommitMessageSerializerTest.java
paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/AppendOnlyTableITCase.java

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/append/CompactionTask.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+import java.util.Objects;
+
+import static org.apache.paimon.utils.SerializationUtils.deserializedBytes;
+import static org.apache.paimon.utils.SerializationUtils.serializeBytes;
+
+/**
+ * Compaction task generated by {@link AppendOnlyTableCompactionCoordinator} and executed be {@link
+ * AppendOnlyTableCompactionWorker}.
+ */
+public class CompactionTask implements CommitMessage {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/CompactionTask.java:47]: It does not need to implement `CommitMessage`.

**Review by JingsongLi — COMMENTED** [review]
We can document this mode is `append-only non-bucket mode`.
All comments.

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCheckUtils.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.paimon.append;
+
+import org.apache.paimon.table.AppendOnlyFileStoreTable;
+import org.apache.paimon.table.Table;
+
+/** Utils for {@link AppendOnlyFileStoreTable}. */
+public class AppendOnlyTableCheckUtils {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCheckUtils.java:25]: remove this class.

File: paimon-core/src/main/java/org/apache/paimon/table/AppendOnlyFileStoreTable.java [hunk 3]
```diff
@@ -141,4 +160,16 @@ public TableWriteImpl<InternalRow> newWrite(
                     return record.row();
                 });
     }
+
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/AppendOnlyFileStoreTable.java:163]:
```
public NonBucketCompaction createNonBucketCompaction() {
     1. check bucket is -1
     2. create NonBucketCompaction
}

class NonBucketCompaction {
     1. createCoordinator
     2. createWorker
}
```

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        this.minFileNum = minFileNum;
+        this.maxFileNum = maxFileNum;
+        snapshotId = snapshotManager.latestSnapshotId();
+
+        if (snapshotId != null) {
+            updateFull();
+        }
+    }
+
+    public void updateRestore() {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java:69]: I cannot image how to use these methods.

File: paimon-core/src/main/java/org/apache/paimon/append/CompactionTask.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    private static final long serialVersionUID = 1L;
+
+    private static final ThreadLocal<CommitMessageSerializer> CACHE =
+            ThreadLocal.withInitial(CommitMessageSerializer::new);
+
+    private BinaryRow partition;
+    private transient List<DataFileMeta> newFiles;
+    private transient List<DataFileMeta> restoredFiles;
+    private transient List<DataFileMeta> compactBefore;
+    private transient List<DataFileMeta> compactAfter;
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/CompactionTask.java:58]: What is used?

File: paimon-core/src/main/java/org/apache/paimon/table/AppendOnlyFileStoreTable.java [hunk 6]
```diff
@@ -130,9 +135,23 @@ public TableWriteImpl<InternalRow> newWrite(String commitUser) {
...
+            writer.fromEmptyWriter();
+        }
         return new TableWriteImpl<>(
-                store().newWrite(commitUser, manifestFilter),
-                new InternalRowKeyAndBucketExtractor(tableSchema),
+                writer,
+                new InternalRowKeyAndBucketExtractor(tableSchema) {
+                    @Override
+                    public int bucket() {
+                        if (negativeBucket) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/AppendOnlyFileStoreTable.java:149]: We should have a `BucketProcessor` to deal with all bucket related things for source and sink.
You can finish this in next PR.

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+                                                v.stream()
+                                                        .map(ManifestEntry::file)
+                                                        .filter(
+                                                                dataFileMeta ->
+                                                                        dataFileMeta.fileSize()
+                                                                                < targetFileSize)
+                                                        .collect(Collectors.toList())));
+    }
+
+    private void updateDelta(Long targetSnapshotId) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java:101]: We don't need to `updateDelta`, just update full and store delta files from writers.

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionWorker.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+/** {@link AppendOnlyFileStoreTable} compact worker. */
+public class AppendOnlyTableCompactionWorker {
+    private final List<CompactionTask> tasks = new ArrayList<>();
+    private final List<CommitMessage> result = new ArrayList<>();
+    private final String commitUser = UUID.randomUUID().toString();
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionWorker.java:36]: Don't create random user, just pass user into this class

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionWorker.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        this.tasks.addAll(tasks);
+        return this;
+    }
+
+    public AppendOnlyTableCompactionWorker accept(CompactionTask task) {
+        this.tasks.add(task);
+        return this;
+    }
+
+    public AppendOnlyTableCompactionWorker doCompact() throws Exception {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionWorker.java:55]: return `List<CommitMessage>`.

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+    private Long snapshotId;
+    private final long targetFileSize;
+    private final int minFileNum;
+    private final int maxFileNum;
+
+    private final Map<BinaryRow, PartitionCompactCoordinator> partitionCompactCoordinators =
+            new HashMap<>();
+
+    public AppendOnlyTableCompactionCoordinator(
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java:52]: Can you modify the `CompactAction` for this non-bucket append-only table?
You can modify topology like `CompactorSourceBuilder`.

**Comment by leaves12138** [comment]
will be apart be more than one pull request
